semtypes/src/main/java/io/ballerina/types/subtypedata/XmlSubtype.java

```diff
@@ -0,0 +1,137 @@
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package io.ballerina.types.subtypedata;
+
+import io.ballerina.types.Bdd;
+import io.ballerina.types.ComplexSemType;
+import io.ballerina.types.Conjunction;
+import io.ballerina.types.Core;
+import io.ballerina.types.PredefinedType;
+import io.ballerina.types.ProperSubtypeData;
+import io.ballerina.types.RecAtom;
+import io.ballerina.types.SemType;
+import io.ballerina.types.SubtypeData;
+import io.ballerina.types.UniformSubtype;
+import io.ballerina.types.UniformTypeBitSet;
+import io.ballerina.types.UniformTypeCode;
+import io.ballerina.types.XmlPrimitive;
+import io.ballerina.types.typeops.BddCommonOps;
+
+/**
+ * Implementation specific to basic type xml.
+ *
+ * @since 3.0.0
+ */
+public class XmlSubtype implements ProperSubtypeData {
+    public int primitives;
+    public Bdd sequence;
+
+    private XmlSubtype(int primitives, Bdd sequence) {
+        this.primitives = primitives;
+        this.sequence = sequence;
+    }
+
+    public static XmlSubtype from(int primitives, Bdd sequence) {
+        return new XmlSubtype(primitives, sequence);
+    }
+
+    public static SemType xmlSingleton(int primitives) {
+        return createXmlSemtype(
+                createXmlSubtype(true, primitives, BddAllOrNothing.bddNothing()),
+                createXmlSubtype(false, primitives, BddAllOrNothing.bddNothing())
+        );
+    }
+
+    public static SemType xmlSequence(SemType constituentType) {
+        if (constituentType == PredefinedType.NEVER) {
+            return xmlSequence(xmlSingleton(XmlPrimitive.XML_PRIMITIVE_NEVER));
+        }
+        if (constituentType instanceof UniformTypeBitSet) {
+            return constituentType;
+        } else {
+            ComplexSemType cct = (ComplexSemType) constituentType; 
+            SubtypeData ro = Core.getComplexSubtypeData(cct, UniformTypeCode.UT_XML_RO);
+            ro = (ro instanceof AllOrNothingSubtype) ? ro : makeSequence(true, (XmlSubtype) ro);
+
+            SubtypeData rw = Core.getComplexSubtypeData(cct, UniformTypeCode.UT_XML_RW);
+            rw = (rw instanceof AllOrNothingSubtype) ? rw : makeSequence(false, (XmlSubtype) rw);
+
+            return createXmlSemtype(ro, rw);
+        }
+    }
+
+    private static SubtypeData makeSequence(boolean roPart, XmlSubtype d) {
+        int primitives = XmlPrimitive.XML_PRIMITIVE_NEVER | d.primitives;
+        int atom = d.primitives &
+                (roPart ? XmlPrimitive.XML_PRIMITIVE_RO_SINGLETON : XmlPrimitive.XML_PRIMITIVE_SINGLETON);
+        Bdd sequence = BddCommonOps.bddUnion(BddCommonOps.bddAtom(RecAtom.createRecAtom(atom)), d.sequence);
+        return createXmlSubtype(roPart, primitives, sequence);
+    }
+
+    public static ComplexSemType createXmlSemtype(SubtypeData ro, SubtypeData rw) {
+        if (ro instanceof AllOrNothingSubtype && ((AllOrNothingSubtype) ro).isNothingSubtype()) {
```

Comment:
Shall we extract a private function called isNothingSubtype.

Reply:
And maybe also a isAllSubtype function. 
Also please create an issue to refactor rest of the code to use those functions. We probably use this in many places.

Reply:
Added in #33933